File: go/pkg/embed/openai.go
package embed

import (
	"context"
	"fmt"
	"net/http"

	"github.com/openai/openai-go"
	"github.com/openai/openai-go/option"
)

// OpenAI embedding models.
const (
	// ModelOpenAI3Small is the small embedding model (1536 dims, customizable).
	ModelOpenAI3Small = "text-embedding-3-small"

	// ModelOpenAI3Large is the large embedding model (3072 dims, customizable).
	ModelOpenAI3Large = "text-embedding-3-large"

	// ModelOpenAIAda002 is the legacy model (1536 dims, fixed).
	ModelOpenAIAda002 = "text-embedding-ada-002"
)

const (
	openAIMaxBatch     = 2048 // OpenAI supports up to 2048 inputs per request
	openAIDefaultDim   = 1536
	openAIDefaultModel = ModelOpenAI3Small
)

// OpenAI implements [Embedder] using the OpenAI embeddings API.
//
// This can also be used with any OpenAI-compatible provider (e.g. SiliconFlow)
// by setting WithBaseURL.
type OpenAI struct {
	client *openai.Client
	model  string
	dim    int
}

var _ Embedder = (*OpenAI)(nil)

// NewOpenAI creates an OpenAI embedder.
//
// The apiKey is required and can be obtained from:
// https://platform.openai.com/api-keys
func NewOpenAI(apiKey string, opts ...Option) *OpenAI {
	cfg := config{
		model:      openAIDefaultModel,
		dim:        openAIDefaultDim,
		httpClient: http.DefaultClient,
	}
	for _, o := range opts {
		o(&cfg)
	}

	clientOpts := []option.RequestOption{
		option.WithAPIKey(apiKey),
		option.WithHTTPClient(cfg.httpClient),
	}
	if cfg.baseURL != "" {
		clientOpts = append(clientOpts, option.WithBaseURL(cfg.baseURL))
	}
	client := openai.NewClient(clientOpts...)

	return &OpenAI{
		client: &client,
		model:  cfg.model,
		dim:    cfg.dim,
	}
}

// Embed returns the embedding for a single text.
func (o *OpenAI) Embed(ctx context.Context, text string) ([]float32, error) {
	if text == "" {
		return nil, ErrEmptyInput
	}
	vecs, err := o.EmbedBatch(ctx, []string{text})
	if err != nil {
		return nil, err
	}
	return vecs[0], nil
}

// EmbedBatch returns embeddings for multiple texts.
// Batches larger than 2048 are automatically split into multiple API calls.
func (o *OpenAI) EmbedBatch(ctx context.Context, texts []string) ([][]float32, error) {
	if len(texts) == 0 {
		return nil, ErrEmptyInput
	}

	result := make([][]float32, len(texts))
	for i := 0; i < len(texts); i += openAIMaxBatch {
		end := min(i+openAIMaxBatch, len(texts))
		batch := texts[i:end]

		vecs, err := o.callAPI(ctx, batch)
		if err != nil {
			return nil, fmt.Errorf("embed batch [%d:%d]: %w", i, end, err)
		}
		copy(result[i:], vecs)
	}
	return result, nil
}

// Dimension returns the configured vector dimensionality.
func (o *OpenAI) Dimension() int {
	return o.dim
}

// Model returns the OpenAI model identifier (e.g., "text-embedding-3-small").
func (o *OpenAI) Model() string {
	return o.model
}

func (o *OpenAI) callAPI(ctx context.Context, texts []string) ([][]float32, error) {
	params := openai.EmbeddingNewParams{
		Model:          o.model,
		Input:          openai.EmbeddingNewParamsInputUnion{OfArrayOfStrings: texts},
		Dimensions:     openai.Int(int64(o.dim)),
		EncodingFormat: openai.EmbeddingNewParamsEncodingFormatFloat,
	}

	resp, err := o.client.Embeddings.New(ctx, params)
	if err != nil {
		return nil, err
	}

	vecs := make([][]float32, len(texts))
	for _, item := range resp.Data {
		idx := item.Index
		if idx < 0 || idx >= int64(len(texts)) {
			return nil, fmt.Errorf("unexpected embedding index %d for batch size %d", idx, len(texts))
		}
		vecs[idx] = float64sToFloat32s(item.Embedding)
	}

	// Verify all slots are filled.
	for i, v := range vecs {
		if v == nil {
			return nil, fmt.Errorf("missing embedding for index %d", i)
		}
	}
	return vecs, nil
}

File: rust/embed/src/openai.rs
use reqwest::Client;

use crate::config::EmbedConfig;
use crate::embed::Embedder;
use crate::error::EmbedError;

/// OpenAI embedding models.
pub const MODEL_OPENAI_3_SMALL: &str = "text-embedding-3-small";
pub const MODEL_OPENAI_3_LARGE: &str = "text-embedding-3-large";
pub const MODEL_OPENAI_ADA_002: &str = "text-embedding-ada-002";

const OPENAI_BASE_URL: &str = "https://api.openai.com/v1";
const OPENAI_MAX_BATCH: usize = 2048;
const OPENAI_DEFAULT_DIM: usize = 1536;

/// OpenAI embedder using the OpenAI embeddings API.
///
/// Also works with any OpenAI-compatible provider (e.g. SiliconFlow)
/// via `with_config` and `EmbedConfig::with_base_url`.
pub struct OpenAI {
    client: Client,
    api_key: String,
    model: String,
    dim: usize,
    base_url: String,
}

impl OpenAI {
    pub fn new(api_key: &str) -> Self {
        Self {
            client: Client::new(),
            api_key: api_key.to_string(),
            model: MODEL_OPENAI_3_SMALL.to_string(),
            dim: OPENAI_DEFAULT_DIM,
            base_url: OPENAI_BASE_URL.to_string(),
        }
    }

    pub fn with_config(api_key: &str, cfg: EmbedConfig) -> Self {
        Self {
            client: Client::new(),
            api_key: api_key.to_string(),
            model: if cfg.model.is_empty() {
                MODEL_OPENAI_3_SMALL.to_string()
            } else {
                cfg.model
            },
            dim: if cfg.dimension == 0 {
                OPENAI_DEFAULT_DIM
            } else {
                cfg.dimension
            },
            base_url: if cfg.base_url.is_empty() {
                OPENAI_BASE_URL.to_string()
            } else {
                cfg.base_url
            },
        }
    }

    async fn call_api(&self, texts: &[&str]) -> Result<Vec<Vec<f32>>, EmbedError> {
        crate::openai_compat::call_embedding_api(
            &self.client,
            &self.api_key,
            &self.base_url,
            &self.model,
            self.dim,
            texts,
        )
        .await
    }
}

#[async_trait::async_trait]
impl Embedder for OpenAI {
    async fn embed(&self, text: &str) -> Result<Vec<f32>, EmbedError> {
        if text.is_empty() {
            return Err(EmbedError::EmptyInput);
        }
        let vecs = self.embed_batch(&[text]).await?;
        Ok(vecs.into_iter().next().unwrap())
    }

    async fn embed_batch(&self, texts: &[&str]) -> Result<Vec<Vec<f32>>, EmbedError> {
        if texts.is_empty() {
            return Err(EmbedError::EmptyInput);
        }

        let mut result = Vec::with_capacity(texts.len());
        for chunk in texts.chunks(OPENAI_MAX_BATCH) {
            let vecs = self.call_api(chunk).await?;
            result.extend(vecs);
        }
        Ok(result)
    }

    fn dimension(&self) -> usize {
        self.dim
    }
}
